core/modules/parser/MySqlListener.cc

```diff
@@ -0,0 +1,1328 @@
+    : Adapter(parent, ctx), _uidContext(ctx) {}
+
+    void onExit() override {
+        LOGS(_log, LOG_LVL_DEBUG, __FUNCTION__);
+        // Fetching the string from a Uid shortcuts a large part of the syntax tree defined under Uid
+        // (see MySqlParser.g4). If Adapters for any nodes in the tree below Uid are implemented then
+        // it will have to be handled and this shortcut may not be taken.
+        lockedParent<UidCBH>()->handleUidString(_uidContext->getText());
+    }
+
+private:
+    MySqlParser::UidContext* _uidContext;
+};
+
+
+class ConstantAdapter : public Adapter {
+public:
+    ConstantAdapter(shared_ptr<ConstantCBH> parent, MySqlParser::ConstantContext* ctx)
+    : Adapter(parent, ctx), _constantContext(ctx) {}
+
+    void onExit() override {
+        lockedParent<ConstantCBH>()->handleConstant(_constantContext->getText());
+    }
+
+private:
+    MySqlParser::ConstantContext* _constantContext;
+};
+
+
+/// MySqlListener impl
+
+
+MySqlListener::MySqlListener() {
+    _rootAdapter = std::make_shared<RootAdapter>();
+    _adapterStack.push(_rootAdapter);
+}
+
+
+std::shared_ptr<query::SelectStmt> MySqlListener::getSelectStatement() const {
+    return _rootAdapter->getSelectStatement();
+}
+
+
+// Create and push an Adapter onto the context stack, using the current top of the stack as a callback handler
+// for the new Adapter. Returns the new Adapter.
+template<typename ParentCBH, typename ChildAdapter, typename Context>
+std::shared_ptr<ChildAdapter> MySqlListener::pushAdapterStack(Context * ctx) {
+    auto p = std::dynamic_pointer_cast<ParentCBH>(_adapterStack.top());
+    if (nullptr == p) {
+        int status;
+        std::ostringstream msg;
+        msg << "can't acquire expected Adapter " <<
+                abi::__cxa_demangle(typeid(ParentCBH).name(),0,0,&status) <<
+                " from top of listenerStack.";
+        LOGS(_log, LOG_LVL_ERROR, msg.str());
+        throw adapter_order_error(msg.str());
+    }
+    auto childAdapter = std::make_shared<ChildAdapter>(p, ctx);
+    _adapterStack.push(childAdapter);
+    childAdapter->onEnter();
+    return childAdapter;
+}
+
+
+template<typename ChildAdapter>
+void MySqlListener::popAdapterStack() {
+    shared_ptr<Adapter> adapterPtr = _adapterStack.top();
+    adapterPtr->onExit();
+    _adapterStack.pop();
+    shared_ptr<ChildAdapter> derivedPtr = dynamic_pointer_cast<ChildAdapter>(adapterPtr);
+    if (nullptr == derivedPtr) {
+        int status;
+        LOGS(_log, LOG_LVL_ERROR, "Top of listenerStack was not of expected type. " <<
+                "Expected: " << abi::__cxa_demangle(typeid(ChildAdapter).name(),0,0,&status) <<
+                " Actual: " << abi::__cxa_demangle(typeid(adapterPtr).name(),0,0,&status) <<
+                " Are there out of order or unhandled listener exits?");
+        // might want to throw here...?
+    }
+}
+
+
+// might want to use this in popAdapterStack?
```

Comment:
This method does not seem to be used anywhere?

Reply:
✅